Framework/Geometry/src/Instrument/StructuredDetector.cpp

```diff
@@ -0,0 +1,684 @@
+            double xrb, double ylb, double ylf, double yrf, double yrb) {
+
+  std::ostringstream shapestr;
+
+  // Create XML shape used to describe detector pixel
+  shapestr << "<type name=\"userShape\" >";
+  shapestr << "<hexahedron id=\"" << name << "\" >";
+  shapestr << "<left-back-bottom-point x=\"" << xlb << "\""
+           << " y=\"" << ylb << "\""
+           << " z=\"0\" />";
+  shapestr << "<left-front-bottom-point x=\"" << xlf << "\""
+           << " y=\"" << ylf << "\""
+           << " z=\"0\" />";
+  shapestr << "<right-front-bottom-point x=\"" << xrf << "\""
+           << " y=\"" << yrf << "\""
+           << " z=\"0\" />";
+  shapestr << "<right-back-bottom-point x=\"" << xrb << "\""
+           << " y=\"" << yrb << "\""
+           << " z=\"0\" />";
+  shapestr << "<left-back-top-point x=\"" << xlb << "\""
+           << " y=\"" << ylb << "\""
+           << " z=\"0.001\" />";
+  shapestr << "<left-front-top-point x=\"" << xlf << "\""
+           << " y=\"" << ylf << "\""
+           << " z=\"0.001\" />";
+  shapestr << "<right-front-top-point x=\"" << xrf << "\""
+           << " y=\"" << yrf << "\""
+           << " z=\"0.001\" />";
+  shapestr << "<right-back-top-point x=\"" << xrb << "\""
+           << " y=\"" << yrb << "\""
+           << " z=\"0.001\" />";
+  shapestr << "</hexahedron>";
+  shapestr << "<bounding-box>";
+  shapestr << "<x-min val=\"" << std::min(xlf, xlb) << "\" />";
+  shapestr << "<x-max val=\"" << std::max(xrb, xrf) << "\" />";
+  shapestr << "<y-min val=\"" << ylb << "\" />";
+  shapestr << "<y-max val=\"" << yrf << "\" />";
+  shapestr << "<z-min val=\"0\" />";
+  shapestr << "<z-max val=\"0.001\" />";
+  shapestr << "</bounding-box>";
+  shapestr << "<algebra val=\"" << name << "\" />";
+  shapestr << "</type>";
+
+  Mantid::Geometry::ShapeFactory shapeCreator;
+
+  return shapeCreator.createShape(shapestr.str(), false);
+}
+
+/** Creates new hexahedral detector pixel at row x column y using the
+*   detector vertex values.
+* @param parent :: The parent component assembly
+* @param name :: The pixel name identifier
+* @param x :: The pixel row
+* @param y :: The pixel column
+* @param id :: The pixel ID
+* @return newly created detector.
+*/
+Detector *StructuredDetector::addDetector(CompAssembly *parent,
+                                          const std::string &name, int x, int y,
+                                          int id) {
+  auto w = m_xpixels + 1;
+
+  // Store hexahedral vertices for detector shape
+  auto xlb = m_xvalues[(y * w) + x];
+  auto xlf = m_xvalues[(y * w) + x + w];
+  auto xrf = m_xvalues[(y * w) + x + w + 1];
+  auto xrb = m_xvalues[(y * w) + x + 1];
+  auto ylb = m_yvalues[(y * w) + x];
+  auto ylf = m_yvalues[(y * w) + x + w];
+  auto yrf = m_yvalues[(y * w) + x + w + 1];
+  auto yrb = m_yvalues[(y * w) + x + 1];
+
+  // calculate midpoint of trapeziod
+  double a = std::abs((double)(xrf - xlf));
```

Comment:
I did this because the system test kept failing at this point even though xrf and xlf are implicitly double. For some reason they std::abs was using the integer version of the method. (Not sure why). I kept getting a warning about converting int to double. It only stopped when I did what you see there.